src/db/couchUsers.ts

```diff
@@ -0,0 +1,183 @@
+    throw error
+  })
+  const user = asCouchUser(raw)
+
+  const currencyCodesToUnregister = Object.keys(
+    user.doc.notifications.currencyCodes
+  ).filter(code => !currencyCodes.includes(code))
+  for (const code of currencyCodesToUnregister) {
+    delete user.doc.notifications.currencyCodes[code]
+  }
+
+  for (const code of currencyCodes) {
+    if (code in user.doc.notifications.currencyCodes) continue
+
+    user.doc.notifications.currencyCodes[code] = {
+      '1': true,
+      '24': true
+    }
+  }
+  await saveToDB(db, user, asCouchUser)
+}
+export const devicesByCurrencyHours = async (
+  hours: string,
+  currencyCode: string,
+  connection: ServerScope
+) => {
+  return await connection.db
+    .use(usersSetup.name)
+    .view<IDevicesByCurrencyHoursViewResponse>('filter', 'by-currency', {
+      key: [currencyCode, hours]
+    })
+}
+
+export const attachDevice = async (
+  userId: string,
+  deviceId: string,
+  connection: ServerScope,
+  userDoc?: CouchUser
+): Promise<void> => {
+  const db = connection.db.use(usersSetup.name)
+  const deviceConnection = connection.db.use(devicesSetup.name)
+  const raw = await deviceConnection.get(deviceId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const device = asCouchDevice(raw)
+  if (unpackDevice(device) == null)
+    throw new Error('Device must be registered before attaching to user.')
+  if (userDoc == null) {
+    const raw = await db.get(userId).catch(error => {
+      if (asMaybeNotFoundError(error) != null) return
+      throw error
+    })
+    userDoc = asCouchUser(raw)
+  }
+  userDoc.doc.devices[deviceId] = true
+  await saveToDB(db, userDoc, asCouchUser)
+}
+
+export const saveUserToDB = async (
+  user: User,
+  connection: ServerScope
+): Promise<void> => {
+  const db = connection.db.use(usersSetup.name)
+  const raw = await db.get(user.userId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const userDoc = asCouchUser(raw)
+  userDoc.doc = user
+  await saveToDB(db, userDoc, asCouchUser)
+}
+
+export const fetchDevices = async (
+  userId: string,
+  connection: ServerScope
+): Promise<Device[]> => {
+  const db = connection.db.use(usersSetup.name)
+  const deviceConnection = connection.db.use(devicesSetup.name)
+  const raw = await db.get(userId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const userDoc = asCouchUser(raw)
+  const devices: Device[] = []
+
+  let updated = false
+  for (const deviceId in userDoc.doc.devices) {
+    const raw = await deviceConnection.get(deviceId).catch(error => {
+      if (asMaybeNotFoundError(error) != null) return
+      throw error
+    })
+    const deviceDoc = asCouchDevice(raw)
+    const device = unpackDevice(deviceDoc)
+    if (device != null) {
+      devices.push(device)
+      continue
+    }
+
+    delete userDoc.doc.devices[deviceId]
+    updated = true
+  }
```

Comment:
I think this query should go into the devices file, something like `getDevicesByUser(connection, user: User): Promise<Device[]>`. I am also unhappy with the idea that a query is deleting stuff from the database - this seems really messed up.

Maybe there should be a `cleanUpMissingDevices` helper that can be run to perform the deletion, so it's:

```js
import {getDevicesByUser} from 'couchDevices'
import {cleanUpMissingDevices} from 'couchUsers'

const devices = await getDevicesByUser(connection, user)
await cleanUpMissingDevices(connection, user, devices)
```

That makes it clearer what's going on.

Reply:
This method no longer exists per your last comment. The logics were moved directly to the HTTP view.